First proposal to "remove" single packet number space option

mirjak opened the pull request:
This PR removes the use of a single packet number space on multiple paths as a fully supported option. However, it currently doesn't remove it entirely.
Specifically, it moves the respective text for now into the appendix but still allows the reception of packets without a Connection ID on multiple validated paths, similarly as path migration is not strictly forbidden without connection ID.

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -1324,3 +1193,141 @@ one of the original proposals are:
+
+--- back
+
+# Sending packet on multiple paths with zero length connection ID
```
```suggestion
# Sending packets on multiple paths with zero length connection ID
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -1324,3 +1193,141 @@ one of the original proposals are:
+delivery on loss detection, congestion control, and ECN handling.
+This proposal specifies algorithms for
+controlling the size of acknowledgement packets and ECN handling in
+Section {{using-zero-length}} and {{ecn-handling}}.
```
```suggestion
{{ecn-handling}}.
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -663,162 +654,39 @@ the Destination Connection ID present in the packet header, as defined
-paths. In that case:
+Further, if the initial path uses zero-length Connection ID and
+packets without a Connection ID are successfully received over another validated path,
+these packet MUST either be acknowledged using ACK_MP frames with packet number space ID 0,
```
```suggestion
these packets MUST either be acknowledged using ACK_MP frames with packet number space ID 0,
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -663,162 +654,39 @@ the Destination Connection ID present in the packet header, as defined
-will improve, but load sharing between multiple paths will not work.
-
-## Using non-zero length CID and Multiple Packet Number Spaces
+However, even if packet without a Connection ID are successfully received
```
```suggestion
However, even if packets without a Connection ID are successfully received
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -663,162 +654,39 @@ the Destination Connection ID present in the packet header, as defined
+However, even if packets without a Connection ID are successfully received
+on multiple paths, it is RECOMMENDED to only use one of the active paths
+for sending of ack-elicting packets with zero-length Connection ID as this
+avoids large ACK frames and ambiguity if loss detection and RTT estimation
```
```suggestion
avoids large ACK frames and ambiguity in loss detection and RTT estimation
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -663,162 +654,39 @@ the Destination Connection ID present in the packet header, as defined
+a Connection ID is received on a 5-tuple of a non validated path,
+the receiver may try to decrypt the packet assuming the packet number
+space of the initial path with Path ID 0. This usually is only plausible
+if only the source address or port are different or if a specific 
```
```suggestion
if only the source address or port are different or if a specific
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -873,7 +741,8 @@ In order to guarantee the uniqueness of the nonce, the nonce N is
 and with the path identifier.
 
-The path ID for 1-RTT packets is the sequence number of {{QUIC-TRANSPORT}}, or
+The path ID for 1-RTT packets is the sequence number of the Connection ID 
```
```suggestion
The path ID for 1-RTT packets is the sequence number of the Connection ID
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -1324,3 +1193,141 @@ one of the original proposals are:
+   operation, so that a series of packets sent from a single path
+   uses a series of consecutive sequence numbers without creating
+   holes.
+ 
```
```suggestion

```

huitema reviewed (commented):
I don't think this is ready yet. We may need to agree on the broad lines first.

mirjak commented:
@huitema yes I also opened issue #152 for discussion!

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -846,15 +696,15 @@ the server provided N Connection IDs, and the client is already actively
-different path, or the same path identified by the Path Identifier, based on
-different strategies of sending ACK_MP frames.
+ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be send on the same path 
+as identified by the Path Identifier. However, an endpoint can be returned via either a
```
```suggestion
as identified by the Path Identifier. However, an ACK_MPfFrame can be returned via either a
```

mirjak commented:
I updated the PR and removed the appendix for SPNP. Further this PR threats the use of the multipath extension without CID as an error new text is the following:

```
If the parameter is set to 1, both endpoints MUST use non-zero connection
IDs. If an enable_multipath parameter set to 1 is received and the carrying packet
does not contain a non-zero legnth connection, the receiver MUST treat this as a connection error of type
TRANSPORT_PARAMETER_ERROR (specified in {{Section 20.1 of QUIC-TRANSPORT}})
and close the connection.
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -846,15 +696,15 @@ the server provided N Connection IDs, and the client is already actively
-ACK_MP frame (defined in {{ack-mp-frame}}) can be returned via either a
-different path, or the same path identified by the Path Identifier, based on
-different strategies of sending ACK_MP frames.
+ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be send on the same path 
```
```suggestion
ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be send on the same path
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -846,15 +696,15 @@ the server provided N Connection IDs, and the client is already actively
-different path, or the same path identified by the Path Identifier, based on
-different strategies of sending ACK_MP frames.
+ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be send on the same path
+as identified by the Path Identifier. However, an ACK_MP frame can be returned via either a
+different path, based on different strategies of sending ACK_MP frames.
```
```suggestion
as identified by the Path Identifier. However, an ACK_MP frame can be returned via a
different path, based on different strategies of sending ACK_MP frames.
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -176,12 +172,11 @@ We assume that the reader is familiar with the terminology used in
   the destination Connection ID used for sending packets on that
-  particular path, but alternative definitions can be used if the length
-  of that connection ID is zero.
+  particular path. 
```
```suggestion
  particular path.
```

mirjak commented on draft-ietf-quic-multipath.md:
```diff
@@ -250,6 +242,12 @@ the endpoints MUST fallback to
 
+If the parameter is set to 1, both endpoints MUST use non-zero connection
+IDs. If an enable_multipath parameter set to 1 is received and the carrying packet
+does not contain a non-zero legnth connection, the receiver MUST treat this as a connection error of type
```
```suggestion
does not contain a non-zero length connection ID, the receiver MUST treat this as a connection error of type
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -104,9 +104,14 @@ QUIC version 1)
 address as well as source and destination port. Therefore, there can be
 at most one active paths/connection ID per 4-tuple.
+  * If the 4-tuple changes without the use of a new connection ID (e.g.
+due to a NAT rebinding), this is considered as a migration events and not
+an new path.
```
```suggestion
due to a NAT rebinding), this is considered as a migration event and not
as a new path.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -192,10 +187,7 @@ Identifier, is that the Path Identifier is used in multipath control
 spaces for different paths. Both identifiers have the same value, which
-is the sequence number of the connection ID, if a non-zero connection ID
-is used. If the connection ID is zero length, the Packet Number Space
-Identifier is 0, while the Path Identifier is selected on path
-establishment.
+is the sequence number of the connection ID, when a new path is established.
```
Maybe it can be addressed later, but given that the Path Identifier and the PNS ID are now always the same, we may simplify this section and only introduce one concept (i.e., the Path Identifier).

mirjak replied:
Yes, I was planning to raise a separate issue and PR after this one is merged. I think that would be cleaner.

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -250,6 +242,12 @@ the endpoints MUST fallback to
 {{QUIC-TRANSPORT}} with single active path and MUST NOT use any frame or
 mechanism defined in this document.
```
Given that we drop support for single packet number space and zero-length CID, we need to change the value of `enable_multipath`. Running a random generator gave me `0x0f739bbc1b666db9`, but we can of course choose another value (for instance, `0x0f739bbc1b666dXX` where `XX` is the draft version; the next version 04 of this document could use  `0x0f739bbc1b666d04`). The IANA table should be updated as well.

mirjak replied:
Yes, I agree that we need to change. Just need to decide with number.

qdeconinck replied:
Just committed 13e6610576 to address this.

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -676,184 +671,8 @@ ACK frames are received in 1-RTT packets while the state of multipath
-If packets contain a non-zero CID, each path has
-its own packet number space for transmitting 1-RTT packets and a new
-ACK frame format is used as specified in {{ack-mp-frame}}.
+A new ACK frame format is used as specified in {{ack-mp-frame}} that
+MUST be used for for acknolowing 1-RTT packets.
```
The added text conflicts with the latter one saying "endpoints SHOULD use ACK_MP frames instead of ACK", and I don't think receiving ACK frames to acknowledge path 0 is actually an issue. Maybe this would be better?
```suggestion
The ACK_MP frame, as specified in {{ack-mp-frame}}, is used to
acknowledge 1-RTT packets.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -863,9 +682,8 @@ Therefore, the packet number space for 1-RTT packets can be identified
-endpoints SHOULD use ACK_MP frames instead of ACK frames for
-acknowledgements of 1-RTT packets on path 0, as well as for 0-RTT packets
-that are acknowledged after the handshake concluded.
+endpoints SHOULD use ACK_MP frames instead of ACK frames also for 0-RTT packets
+that are acknowledged after the handshake concluded using the Packet Number Space Identifier 0.
```
Maybe clearer?
```suggestion
endpoints SHOULD use ACK_MP frames instead of ACK frames to acknowledge application
data packets, including 0-RTT packets, received on path ID 0 after the handshake concluded.
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -877,15 +695,15 @@ the server provided N Connection IDs, and the client is already actively
-different path, or the same path identified by the Path Identifier, based on
-different strategies of sending ACK_MP frames.
+ACK_MP frame (defined in {{ack-mp-frame}}) SHOULD be send on the same path
+as identified by the Path Identifier. However, an ACK_MP frame can be returned via a
```
No strong opinion here, but should the "an ACK_MP frame can be returned" be instead "an ACK_MP frame MAY be returned"?

mirjak replied:
No, the normative part is the active part ("SHOULD send"). This other sentence should not be normative because it's just explaining that this can happen and needs to be considered.

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -904,8 +722,8 @@ In order to guarantee the uniqueness of the nonce, the nonce N is
-The path ID for 1-RTT packets is the sequence number of {{QUIC-TRANSPORT}}, or
-zero if the Connection ID is zero-length.  {{Section 19 of QUIC-TRANSPORT}}
+The path ID for 1-RTT packets is the sequence number of the Connection ID
+as specfied in {{QUIC-TRANSPORT}}.  {{Section 19 of QUIC-TRANSPORT}}
```
```suggestion
as specified in {{QUIC-TRANSPORT}}.  {{Section 19 of QUIC-TRANSPORT}}
```

qdeconinck commented on draft-ietf-quic-multipath.md:
```diff
@@ -1139,14 +920,11 @@ path identifier.
-    This method SHOULD be used if this connection identifier is non-zero
-    length. This method MUST NOT be used if this connection identifier
-    is zero-length.
+    This method SHOULD be used by default.
```
We can of course delegate this on a subsequent PR, but I think we should get rid of the `Identifier Type` for the PATH_ABANDON and PATH_STATUS frames. Also, I wonder it the value 1 should rather be the default (given #137).

mirjak replied:
Yes, I was also planning to raise a separate issue and PR on this one after the merge.

qdeconinck reviewed (changes requested):
Globally good to go, a few nits to fix before merging. Also, given the interoperability changes it involves, I suggest to already change the transport parameter value in this PR (see my specific comment below).

qdeconinck reviewed (approved):
With my added commit (13e661057), LGTM.

huitema reviewed (approved):
Yes, I approve. It looks very much like I remember of my first review of this PR, with nice fixes.
